packages/backend/src/services/warehouseClients/DatabricksWarehouseClient.ts

```diff
@@ -0,0 +1,109 @@
+import odbc, { Result } from 'odbc';
+import { CreateDatabricksCredentials, DimensionType } from 'common';
+import { WarehouseCatalog, WarehouseClient } from '../../types';
+import { WarehouseConnectionError, WarehouseQueryError } from '../../errors';
+
+export const DRIVER_PATH = '/opt/simba/spark/lib/64/libsparkodbc_sb64.so';
+
+type SparkSchemaResult = {
+    TABLE_CAT: string;
+    TABLE_SCHEM: string;
+    TABLE_NAME: string;
+    COLUMN_NAME: string;
+    DATA_TYPE: number;
+    TYPE_NAME: string;
+};
+
+export default class DatabricksWarehouseClient implements WarehouseClient {
+    connectionString: string;
+
+    constructor({
+        serverHostName,
+        port,
+        personalAccessToken,
+        httpPath,
+    }: CreateDatabricksCredentials) {
+        this.connectionString = `Driver=${DRIVER_PATH};Server=${serverHostName};HOST=${serverHostName};PORT=${port};SparkServerType=3;Schema=default;ThriftTransport=2;SSL=1;AuthMech=3;UID=token;PWD=${personalAccessToken};HTTPPath=${httpPath};UseNativeQuery=1`;
+    }
+
+    async runQuery(sql: string): Promise<Record<string, any>[]> {
+        let connection: odbc.Connection;
+        try {
+            connection = await odbc.connect(this.connectionString);
+        } catch (e) {
+            console.error(e);
+            throw new WarehouseConnectionError(e.message);
+        }
+        try {
+            const results = await connection.query<Record<string, any>>(sql);
+            return results;
+        } catch (e) {
+            throw new WarehouseQueryError(e.message);
+        } finally {
+            await connection.close();
+        }
+    }
+
+    async test(): Promise<void> {
+        await this.runQuery('SELECT 1');
+    }
+
+    async getCatalog(
+        requests: {
+            database: string;
+            schema: string;
+            table: string;
+            columns: string[];
+        }[],
+    ) {
+        let pool: odbc.Pool;
+        let results: Result<SparkSchemaResult>[];
+        try {
+            pool = await odbc.pool(this.connectionString);
+        } catch (e) {
+            throw new WarehouseConnectionError(e.message);
+        }
+        try {
+            const promises = requests.map(async (request) => {
+                let connection: odbc.Connection;
+                try {
+                    connection = await pool.connect();
+                } catch (e) {
+                    throw new WarehouseConnectionError(e.message);
+                }
+                try {
+                    const columns = (await connection.columns(
+                        // @ts-ignore
+                        'SPARK', // This is always SPARK
+                        request.schema,
+                        request.table,
+                        '',
+                    )) as Result<SparkSchemaResult>;
+                    return columns;
+                } catch (e) {
+                    throw new WarehouseQueryError(e.message);
+                } finally {
```

Comment:
Do we want to ignore the error if is a 'not found' ? similar to what we have in the bigquery client ?

Reply:
The results will simply be empty if the schema-table combination isn't found.

Although in checking this I've realised that if you specify a schema-table combination that isn't found then you get no columns in the resulting catalog, e.g.:

```javascript
result = {
  SPARK: {
    missingSchema: {
      missingTable: {}
    }
  }
}
```

Is that correct? Do we actually need the `columns` in the `requests` argument?

Reply:
In my snowflake PR you requested that the returned warehouse catalog should not contain columns that were not in the **requests** argument.

Reply:
IIRC, it was that the function returned tables that were not in the `request` argument, I'm not sure how needed it is to request specific columns.

Reply:
Might have misunderstand then. With our current business logic I think is fine either way.